### DISCUSSION TITLE
How to increase density of vehicles in an environment?

### DISCUSSION THREAD

**Discussion by mhtb32** [eleurent/highway-env#168] [Ideas]
I've made an environment like `highway-v0`, which vehicle's action space is continuous, and it tries to reach a goal while avoiding a crash. Here's an episode of a learned agent with the SAC model.

![An episode of the mentioned environment](https://imgur.com/83naNvi.gif)

As the density of vehicles in the environment is low, the agent has not a difficult job avoiding them(and it seems it is the IDM vehicle which is avoiding the agent?). I tried to make it harder for the agent by increasing the number of vehicles, but the density of vehicles around remains the same. How can I make it harder, like the animation in the README file from the `highway-v0` environment?

**Comment by eleurent** [comment]
There has been a changed in the repartition of vehicles, I switched it back to something closer to what it used to be. You can still control the density by changing the `"initial_spacing"` environment config

**Comment by eleurent** [comment]
Nice environment, by the way! :+1:

**Comment by eleurent** [comment]
Oh, and I do not think that other vehicles are trying to avoid the ego-vehicle (unless you set the POLITENESS parameter, but it is 0 by default). However, the 3rd vehicle may be trying to avoid the landmark (I can't remember how we handled it, but I think it is considered as a fixed object rather than ignored, or was it only for obstacles?). For the two first I don't really know, 2nd one might change lane if it started on the same as 1st, but I can't really tell here. MOBIL can sometimes be quite unstable at start, and needs a few seconds to reach an equilibrium.

**Comment by mhtb32** [comment]
> However, the 3rd vehicle may be trying to avoid the landmark (I can't remember how we handled it, but I think it is considered as a fixed object rather than ignored, or was it only for obstacles?).

Yes, they try to avoid the landmark, and the strange thing is sometimes IDM vehicles stop before it. See this, for example, (The ego vehicle's training is incomplete):

![Imgur](https://imgur.com/EM4PTzH.gif)

Such behavior is a blocker to an agent's exploration.

Maybe better behavior is that IDM vehicles ignore landmarks but go around the obstacles.

**Comment by mhtb32** [comment]
Sometimes the IDM's stop before the landmark, and we have a crash :joy:

![crash](https://imgur.com/vdX6jIm.gif)

But sometimes they go around the landmark, and everything's fine:

![success](https://i.imgur.com/r3Tdd4q.gif)

**Comment by eleurent** [comment]
I've seen this behaviour before: once it sees an obstacle, the IDM model starts braking, and this causes the MOBIL model to consider a lane change.

1. If a lane change is possible (it would not cut-in in front of another vehicle), a lane change is triggered and the vehicle avoids the obstacle/landmark
2. Else (would cut-in), the vehicle cannot change lane and can only brake and stop. Once it is stopped at the obstacle, the MOBIL now recommends a lane change, but IDM still outputs an acceleration of 0 (since the obstacle is close), so the vehicle is stuck.
